### PR TITLE
expression: Added type compare check for int/binary strings for BETWEEN, issue-23602

### LINKED ISSUE #23602: Incompatible casting of hexadecimal literals
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)

```
drop table if exists t1;
create table t1(
    id bigint unsigned
);
CREATE INDEX id ON t1(id);
EXPLAIN SELECT * FROM t1 WHERE id between x'e0' and x'e1';
SHOW WARNINGS;
EXPLAIN SELECT * FROM t1 WHERE id between 224 and 225;
SHOW WARNINGS;
```

### 2. What did you expect to see? (Required)

MySQL successfully casts the hexadecimal notation strings to integers and is able to use the index. TiDB fails to cast them correctly and instead performs a table scan.

```
mysql> drop table if exists t1;
Query OK, 0 rows affected (0.003 sec)

mysql> create table t1(
    ->     id bigint unsigned
    -> );
Query OK, 0 rows affected (0.007 sec)

mysql> CREATE INDEX id ON t1(id);
Query OK, 0 rows affected (0.007 sec)

mysql> EXPLAIN SELECT * FROM t1 WHERE id between x'e0' and x'e1';
+----+-------------+-------+------------+-------+---------------+------+---------+------+------+----------+--------------------------+
| id | select_type | table | partitions | type  | possible_keys | key  | key_len | ref  | rows | filtered | Extra                    |
+----+-------------+-------+------------+-------+---------------+------+---------+------+------+----------+--------------------------+
|  1 | SIMPLE      | t1    | NULL       | index | id            | id   | 9       | NULL |    1 |   100.00 | Using where; Using index |
+----+-------------+-------+------------+-------+---------------+------+---------+------+------+----------+--------------------------+
1 row in set, 1 warning (0.000 sec)

mysql> SHOW WARNINGS;
+-------+------+--------------------------------------------------------------------------------------------------------------+
| Level | Code | Message                                                                                                      |
+-------+------+--------------------------------------------------------------------------------------------------------------+
| Note  | 1003 | /* select#1 */ select `test`.`t1`.`id` AS `id` from `test`.`t1` where (`test`.`t1`.`id` between 224 and 225) |
+-------+------+--------------------------------------------------------------------------------------------------------------+
1 row in set (0.000 sec)

mysql> EXPLAIN SELECT * FROM t1 WHERE id between 224 and 225;
+----+-------------+-------+------------+-------+---------------+------+---------+------+------+----------+--------------------------+
| id | select_type | table | partitions | type  | possible_keys | key  | key_len | ref  | rows | filtered | Extra                    |
+----+-------------+-------+------------+-------+---------------+------+---------+------+------+----------+--------------------------+
|  1 | SIMPLE      | t1    | NULL       | index | id            | id   | 9       | NULL |    1 |   100.00 | Using where; Using index |
+----+-------------+-------+------------+-------+---------------+------+---------+------+------+----------+--------------------------+
1 row in set, 1 warning (0.000 sec)

mysql> SHOW WARNINGS;
+-------+------+--------------------------------------------------------------------------------------------------------------+
| Level | Code | Message                                                                                                      |
+-------+------+--------------------------------------------------------------------------------------------------------------+
| Note  | 1003 | /* select#1 */ select `test`.`t1`.`id` AS `id` from `test`.`t1` where (`test`.`t1`.`id` between 224 and 225) |
+-------+------+--------------------------------------------------------------------------------------------------------------+
1 row in set (0.000 sec)
```

### 3. What did you see instead (Required)

```
tidb> drop table if exists t1;
Query OK, 0 rows affected (0.193 sec)

tidb> create table t1(
    ->     id bigint unsigned
    -> );
Query OK, 0 rows affected (0.071 sec)

tidb> CREATE INDEX id ON t1(id);
Query OK, 0 rows affected (2.773 sec)

tidb> EXPLAIN SELECT * FROM t1 WHERE id between x'e0' and x'e1';
+-------------------------+----------+-----------+---------------+------------------------------------------------------------------------------------------------------+
| id                      | estRows  | task      | access object | operator info                                                                                        |
+-------------------------+----------+-----------+---------------+------------------------------------------------------------------------------------------------------+
| TableReader_7           | 8000.00  | root      |               | data:Selection_6                                                                                     |
| └─Selection_6           | 8000.00  | cop[tikv] |               | ge(cast(test.t1.id, double UNSIGNED BINARY), 224), le(cast(test.t1.id, double UNSIGNED BINARY), 225) |
|   └─TableFullScan_5     | 10000.00 | cop[tikv] | table:t1      | keep order:false, stats:pseudo                                                                       |
+-------------------------+----------+-----------+---------------+------------------------------------------------------------------------------------------------------+
3 rows in set (0.001 sec)

tidb> SHOW WARNINGS;
Empty set (0.000 sec)

tidb> EXPLAIN SELECT * FROM t1 WHERE id between 224 and 225;
+------------------------+---------+-----------+------------------------+-------------------------------------------------+
| id                     | estRows | task      | access object          | operator info                                   |
+------------------------+---------+-----------+------------------------+-------------------------------------------------+
SHOW WARNINGS;| IndexReader_6          | 250.00  | root      |                        | index:IndexRangeScan_5                          |
| └─IndexRangeScan_5     | 250.00  | cop[tikv] | table:t1, index:id(id) | range:[224,225], keep order:false, stats:pseudo |
+------------------------+---------+-----------+------------------------+-------------------------------------------------+
2 rows in set (0.001 sec)

tidb> SHOW WARNINGS;
Empty set (0.000 sec)
```

### 4. What is your TiDB version? (Required)

```
tidb_version(): Release Version: v4.0.0-beta.2-2460-g40b92181d-dirty
Edition: Community
Git Commit Hash: 40b92181d092437996760f1f092ddf5b1da24d1b
Git Branch: HEAD
UTC Build Time: 2021-03-25 19:55:14
GoVersion: go1.13.4
Race Enabled: false
TiKV Min Version: v3.0.0-60965b006877ca7234adaced7890d7b029ed1306
Check Table Before Drop: false
1 row in set (0.002 sec)
```

### PR DESCRIPTION
**Pull request by mjonss**
### What problem does this PR solve?

Issue Number: close #23602 
Problem Summary: For BETWEEN, it used DOUBLE to compare between INT and Binary Literals, resulting in not being able to use indexes and could result in wrong results due to different resolution and range.

### What is changed and how it works?
What's Changed:
Special case added for BETWEEN comparison.

How it Works:
Now BETWEEN expression checks if all arguments are integer or binary literals and if so, compares them as integers.

I also considered to transform:
v [NOT] BETWEEN min AND max
into
[NOT] (min <= v AND v <= max)
in the parser directly to simplify the code (no special handling of BETWEEN, but just rely on the more tested normal two way comparison).

But that could result in different types when comparing with min and
max.

### Check List 

Tests 

- [ ] Unit test
- [X] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note 

```release-note
Comparing int and binary literal is done as integer for BETWEEN expressions.
```

### CODE REVIEW

**Review by xhebox — APPROVED** [review]

**Comment by xhebox** [comment]
> 
> 
> /cc @pingcap/exec-reviewers

Members of the team are not updated. Maybe send it in slack or lark. /cc @tiancaiamao

**Comment by guo-shaoge** [comment]
/run-integration-common-test                                                                                                                                                                                                                                                  /run-common-test
/run-integration-copr-test

File: executor/executor_test.go [hunk 1]
```diff
@@ -8999,3 +8999,20 @@ func (s *testSuite) TestIssue25447(c *C) {
...
+	defer tk.MustExec("DROP TABLE t")
+	tk.MustExec("INSERT INTO t VALUES (0),(1),(2),(3),(18446744073709551600),(18446744073709551605),(18446744073709551610),(18446744073709551615)")
+	tk.MustExec("ANALYZE TABLE t")
+	tk.MustQuery(`EXPLAIN FORMAT = 'brief' SELECT a FROM t WHERE a >= 0x1 AND a <= 0x2`).Check(testkit.Rows(
+		"TableReader 2.00 root  data:TableRangeScan]\n" +
+			"[└─TableRangeScan 2.00 cop[tikv] table:t range:[1,2], keep order:false"))
+	tk.MustQuery(`EXPLAIN FORMAT = 'brief' SELECT a FROM t WHERE a BETWEEN 0x1 AND 0x2`).Check(testkit.Rows(
+		"TableReader 2.00 root  data:TableRangeScan]\n" +
+			"[└─TableRangeScan 2.00 cop[tikv] table:t range:[1,2], keep order:false"))
+	tk.MustQuery("SELECT a FROM t WHERE a BETWEEN 0xFFFFFFFFFFFFFFF5 AND X'FFFFFFFFFFFFFFFA'").Check(testkit.Rows("18446744073709551605", "18446744073709551610"))
```
Comment (guo-shaoge) [inline executor/executor_test.go:9017]: drop table if exists t1;
create table t1(c1 bigint unsigned);                                                                                                                                                                                                                                          insert into t1 values(18446744073709551610);
select * from t1 where c1 between 18446744073709551601 and 0xFFFFFFFFFFFFFFFE;

Looks like not compatible with MySQL 8.0.

Comment (mjonss) [reply executor/executor_test.go:9017]: That seems to be a bug in MySQL?
```sql
mysql> drop table if exists t1;
Query OK, 0 rows affected (0,06 sec)

mysql> create table t1(c1 bigint unsigned); insert into t1 values(18446744073709551610);
Query OK, 0 rows affected (0,06 sec)

Query OK, 1 row affected (0,02 sec)

mysql> select * from t1 where c1 between 18446744073709551601 and 0xFFFFFFFFFFFFFFFE;
Empty set (0,00 sec)

mysql> explain select * from t1 where c1 between 18446744073709551601 and 0xFFFFFFFFFFFFFFFE;
+----+-------------+-------+------------+------+---------------+------+---------+------+------+----------+-------------+
| id | select_type | table | partitions | type | possible_keys | key  | key_len | ref  | rows | filtered | Extra       |
+----+-------------+-------+------------+------+---------------+------+---------+------+------+----------+-------------+
|  1 | SIMPLE      | t1    | NULL       | ALL  | NULL          | NULL | NULL    | NULL |    1 |   100.00 | Using where |
+----+-------------+-------+------------+------+---------------+------+---------+------+------+----------+-------------+
1 row in set, 1 warning (0,00 sec)

mysql> show warnings;
+-------+------+------------------------------------------------------------------------------------------------------------------------------------------------------+
| Level | Code | Message                                                                                                                                              |
+-------+------+------------------------------------------------------------------------------------------------------------------------------------------------------+
| Note  | 1003 | /* select#1 */ select `test`.`t1`.`c1` AS `c1` from `test`.`t1` where (`test`.`t1`.`c1` between 18446744073709551601 and 18446744073709551614) |
+-------+------+------------------------------------------------------------------------------------------------------------------------------------------------------+
1 row in set (0,00 sec)

mysql> select version();
+-------------------------+
| version()               |
+-------------------------+
| 8.0.26-0ubuntu0.21.04.3 |
+-------------------------+
1 row in set (0,00 sec)

mysql> /* select#1 */ select `test`.`t1`.`c1` AS `c1` from `test`.`t1` where (`test`.`t1`.`c1` between 18446744073709551601 and 18446744073709551614);
+----------------------+
| c1                   |
+----------------------+
| 18446744073709551610 |
+----------------------+
1 row in set (0,01 sec)
```

I consider TiDB to do the right thing after this patch, and suggest to not be bug compatible with MySQL for this case.
@guo-shaoge what do you think?

**Review by guo-shaoge — APPROVED** [review]

**Pull request merged**